File: java/src/jan/week3/boj/카드게임_박재환.java
package jan.week3.boj;

import java.util.*;
import java.io.*;

public class 카드게임_박재환 {
    static BufferedReader br;
    static StringBuilder sb;
    public static void main(String[] args) throws IOException {
        br = new BufferedReader(new InputStreamReader(System.in));
        sb = new StringBuilder();
        init();
        br.close();
        System.out.println(sb);
    }

    static StringTokenizer st;
    static int cardCount, pickCount, turnCount;
    static int[] cards;
    static boolean[] isUsed;
    static void init() throws IOException {
        st = new StringTokenizer(br.readLine().trim());
        cardCount = Integer.parseInt(st.nextToken());
        pickCount = Integer.parseInt(st.nextToken());
        turnCount = Integer.parseInt(st.nextToken());

        st = new StringTokenizer(br.readLine().trim());
        cards = new int[pickCount];
        isUsed = new boolean[pickCount];
        for(int i=0; i<pickCount; i++) {
            int number = Integer.parseInt(st.nextToken());
            cards[i] = number;
        }
        Arrays.sort(cards);     // 오름차순으로 정렬

        st = new StringTokenizer(br.readLine().trim());
        while(turnCount-- > 0) {
            int cardNumber = Integer.parseInt(st.nextToken());
            int result = findMinCardNumber(cardNumber);

            for(; result<pickCount && isUsed[result]; result++);

            isUsed[result] = true;
            sb.append(cards[result]).append('\n');
        }
    }

    /**
     * target 보다 큰 가장 작은 수
     * => upper bound
     */
    static int findMinCardNumber(int target) {
        int l=0, r=pickCount;

        while(l < r) {
            int mid = l + (r-l)/2;

            if(cards[mid] > target) {
                r = mid;
            } else l = mid+1;
        }

        return l;
    }
}

File: python/src/jan/week3/boj/카드게임_박재환.py
def solution():
    import sys
    input = sys.stdin.readline

    card_count, pick_count, turn_count = map(int, input().split())
    cards = list(map(int, input().split()))

    # Union-Find
    parents = [None]*(pick_count+1)
    def make():
        nonlocal parents
        for i in range(pick_count+1):
            parents[i] = i

    def find(a):
        if parents[a] == a:
            return a;

        parents[a] = find(parents[a])
        return parents[a]

    def union(a, b):
        root_a = find(a)
        root_b = find(b)

        if root_a == root_b:
            return

        parents[root_a] = root_b

    # Binary Search
    def find_upper_bound(target):
        l = 0
        r = pick_count

        while(l < r):
            mid = (l+r) // 2
            if cards[mid] > target:
                r = mid
            else:
                l = mid+1

        return l

    # ----------------------------------

    cards.sort()
    # 1. Union Find 만들기
    make()
    turn = list(map(int, input().split()))
    for card in turn:
        upper_idx = find_upper_bound(card)
        result = find(upper_idx)
        print(cards[result])
        union(result, result+1)

if __name__ == '__main__':
    solution()